server/mdm/apple/apple_mdm.go

```diff
@@ -256,6 +261,105 @@ func (d *DEPService) RegisterProfileWithAppleDEPServer(ctx context.Context, team
 	return requestedTokenProfileUUID, requestedTokenModTime, nil
 }
 
+func (d *DEPService) ValidateSetupAssistant(ctx context.Context, team *fleet.Team, setupAsst *fleet.MDMAppleSetupAssistant, abmTokenOrgName string) error {
+	appCfg, err := d.ds.AppConfig(ctx)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "fetching app config")
+	}
+
+	// must always get the default profile, because the authentication token is
+	// defined on that profile.
+	defaultProf, err := d.ds.GetMDMAppleEnrollmentProfileByType(ctx, fleet.MDMAppleEnrollmentTypeAutomatic)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "fetching default profile")
+	}
+
+	enrollURL, err := EnrollURL(defaultProf.Token, appCfg)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "generating enroll URL")
+	}
+
+	rawJSON := setupAsst.Profile
+
+	var jsonProf godep.Profile
+	jsonProf.IsMDMRemovable = true // the default value defined by Apple is true
+	if err := json.Unmarshal(rawJSON, &jsonProf); err != nil {
+		return ctxerr.Wrap(ctx, err, "unmarshalling DEP profile")
+	}
+
+	// if configuration_web_url is set, this setting is completely managed by the
+	// IT admin.
+	if jsonProf.ConfigurationWebURL == "" {
```

Comment:
from here to the bottom of the function. I'm 100% in agreement with not prematurely abstracting but this logic has so many nuances and it's always shifting (talking by experience here 😓 ) I don't think it's a good idea to duplicate all of this here and in `RegisterProfileWithAppleDEPServer`

Reply:
@roperzh gotcha! 

> I don't think it's a good idea to duplicate all of this here and in RegisterProfileWithAppleDEPServer

I definitely hear this! Would a shared private method that has some of this logic be a better approach? One issue I see (and the main reason I created this new function) is that the logic of "validating" at upload time is different than the "register" logic: there are different errors being returned, and if only 1 ABM token exists we're using that one no matter what.

Reply:
a private method makes sense. The main thing we should abstract (anything else can be duplicated) is how we build the JSON profile, which values we set by default, which ones we override, etc.